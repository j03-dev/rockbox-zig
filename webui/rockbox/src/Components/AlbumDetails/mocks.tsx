export const tracks = [
  {
    id: "1",
    trackNumber: 1,
    title: "Gasoline (Live)",
    artist: "Seether",
    time: "2:57",
  },
  {
    id: "2",
    trackNumber: 2,
    title: "Driven Under (Live)",
    artist: "Seether",
    time: "4:58",
  },
  {
    id: "3",
    trackNumber: 3,
    title: "Diseased (Live)",
    artist: "Seether",
    time: "3:46",
  },
  {
    id: "4",
    trackNumber: 4,
    title: "Truth (Live)",
    artist: "Seether",
    time: "5:15",
  },
  {
    id: "5",
    trackNumber: 5,
    title: "Immortality (Live)",
    artist: "Seether",
    time: "5:02",
  },
  {
    id: "6",
    trackNumber: 6,
    title: "Tied My Hands (Live)",
    artist: "Seether",
    time: "5:16",
  },
  {
    id: "7",
    trackNumber: 7,
    title: "Sympathetic (Live)",
    artist: "Seether",
    time: "4:12",
  },
  {
    id: "8",
    trackNumber: 8,
    title: "Fine Again (Live)",
    artist: "Seether",
    time: "5:05",
  },
  {
    id: "9",
    trackNumber: 9,
    title: "Broken (Live)",
    artist: "Seether",
    time: "4:17",
  },
  {
    id: "10",
    trackNumber: 10,
    title: "The Gift (Live)",
    artist: "Seether",
    time: "5:36",
  },
  {
    id: "11",
    trackNumber: 11,
    title: "Remedy (Live)",
    artist: "Seether",
    time: "3:40",
  },
  {
    id: "12",
    trackNumber: 12,
    title: "Plastci Man (Live)",
    artist: "Seether",
    time: "3:34",
  },
  {
    id: "13",
    trackNumber: 13,
    title: "The Gist (Alternate Mix)",
    artist: "Seether",
    time: "3:24",
  },
];
