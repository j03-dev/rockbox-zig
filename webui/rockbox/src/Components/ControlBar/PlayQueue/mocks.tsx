/* eslint-disable no-loss-of-precision */

export const nextTracks = [
  {
    id: "5ce25a3dadabc01f94130d806294a296",
    title: "Can I",
    artist: "Kodak Black",
    album: "Lil Big Pac",
    cover:
      "https://resources.tidal.com/images/f36fa0ef/fe50/4c54/8b97/3c9f4f72f8e8/320x320.jpg",
    duration: 216083.99963378906,
    artistId: "c38e31502394622bbdfd3d82f6911eb3",
    albumId: "229251493",
  },
  {
    id: "c2754b1edfcb8fb4dd34d94f4e42b622",
    title: "Too Many Years",
    artist: "Kodak Black, PnB Rock",
    album: "Lil Big Pac",
    cover:
      "https://resources.tidal.com/images/f36fa0ef/fe50/4c54/8b97/3c9f4f72f8e8/320x320.jpg",
    duration: 196486.99951171875,
    artistId: "a01003e6ac321328ad1e208bf41206f0",
    albumId: "229251493",
  },
  {
    id: "89ab06746b363d3295a2bc297339c5be",
    title: "ALL THE TIME HIGH",
    artist: "Juicy J featuring Kaash Paige",
    album: "THE HUSTLE STILL CONTINUES (Deluxe)",
    cover:
      "https://resources.tidal.com/images/aeb814c6/7767/421f/a121/46fdc0895c53/320x320.jpg",
    duration: 120141.99829101562,
    artistId: "c96c630a2fd2edd0eec175c4f65e0329",
    albumId: "229251493",
  },
];

export const previousTracks = [
  {
    id: "93b7e61e27b4e5cab4c6ac4d45a4dc9f",
    title: "Sobriety Sucks",
    artist: "Yelawolf",
    album: "Trunk Muzic Returns (Deluxe Edition)",
    cover:
      "https://resources.tidal.com/images/3b8d6b45/b3e4/4f0c/a69d/2f31a1fceadf/320x320.jpg",
    duration: 255001.00708007812,
    artistId: "052dcb0741a91da6d26a35600ec46cd9",
  },
];
