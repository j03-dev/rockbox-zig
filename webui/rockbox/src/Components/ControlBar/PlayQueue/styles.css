.album-cover-container.queue {
  height: 64px;
  justify-content: center;
  align-items: center;
  display: flex;
  position: relative;
  margin-left: 10px;
}

.album-cover-container.queue .floating-play {
  display: none;
  position: absolute;
  left: 8px;
  top: 19px;
 }

.album-cover-container:hover .floating-play {
  display: block;
}

.album-cover-container:hover img {
  opacity: 0.4;
 }