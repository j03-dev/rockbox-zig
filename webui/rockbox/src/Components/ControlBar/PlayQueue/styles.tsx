import styled from "@emotion/styled";
import { LazyLoadImage } from "react-lazy-load-image-component";

export const Container = styled.div`
  height: calc(100vh - 113px);
  width: 370px;
`;

export const Header = styled.div`
  display: flex;
  flex-direction: row;
`;

export const Title = styled.div`
  font-size: 14px;
  margin-left: 16px;
  margin-right: 16px;
  padding-top: 20px;
  margin-bottom: 20px;
  flex: 1;
`;

export const Switch = styled(Title)`
  color: #fe099c;
  flex: 1;
  text-align: end;
  cursor: pointer;
  -webkit-user-select: none;
  -ms-user-select: none;
  user-select: none;
`;

export const List = styled.div`
  height: calc(100% - 59.5px);
  overflow-y: auto;
`;

export const ListItem = styled.div`
  display: flex;
  flex-direction: row;
  height: 64px;
  align-items: center;
  padding-left: 16px;
  padding-right: 16px;
  cursor: pointer;
  &:hover {
    background-color: ${({ theme }) => theme.colors.hover};
  }
`;

export const TrackTitle = styled.div`
  font-size: 14px;
  font-family: RockfordSansMedium;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
`;

export const Artist = styled.div`
  font-size: 14px;
  color: ${({ theme }) => theme.colors.secondaryText};
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
`;

export const TrackDetails = styled.div`
  display: flex;
  min-width: 222px;
  flex-direction: column;
  flex: 1;
`;

export const AlbumCover = styled(LazyLoadImage)<{ current?: boolean }>`
  height: 48px;
  width: 48px;
  border-radius: 4px;
  margin-right: 18px;
  cursor: pointer;
  ${({ current }) => `opacity: ${current ? 0.4 : 1};`}
`;

export const AlbumCoverAlt = styled.div<{ current?: boolean }>`
  height: 48px;
  width: 48px;
  border-radius: 4px;
  cursor: pointer;
  background-color: ${(props) => props.theme.colors.cover};
  display: flex;
  justify-content: center;
  align-items: center;
  margin-right: 18px;
  ${({ current }) => `opacity: ${current ? 0 : 1};`}
`;

export const Remove = styled.button`
  background-color: transparent;
  cursor: pointer;
  border: none;
`;

export const Placeholder = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  height: 100%;
  text-align: center;
  padding-left: 20px;
  padding-right: 20px;
  font-size: 14px;
`;
