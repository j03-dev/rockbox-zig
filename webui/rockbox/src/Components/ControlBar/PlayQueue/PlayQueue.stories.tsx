import type { Meta, StoryObj } from "@storybook/react";

import PlayQueue from "./PlayQueue";
import { fn } from "@storybook/test";
import { nextTracks, previousTracks } from "./mocks";

// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
const meta = {
  title: "Components/PlayQueue",
  component: PlayQueue,
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/api/argtypes
  argTypes: {},
} satisfies Meta<typeof PlayQueue>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/writing-stories/args
export const Default: Story = {
  args: {
    previousTracks,
    nextTracks,
    onPlayTrackAt: fn(),
    onRemoveTrackAt: fn(),
  },
};
