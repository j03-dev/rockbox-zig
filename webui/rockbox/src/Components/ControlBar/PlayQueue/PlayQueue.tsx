import { FC, useMemo, useRef, useState } from "react";
import { Track } from "../../../Types/track";
import { Link } from "react-router-dom";
import { useTheme } from "@emotion/react";
import TrackIcon from "../../Icons/Track";
import { CloseOutline } from "@styled-icons/evaicons-outline";
import { Play } from "@styled-icons/ionicons-sharp";
import {
  AlbumCover,
  AlbumCoverAlt,
  Artist,
  Container,
  Header,
  List,
  ListItem,
  Placeholder,
  Remove,
  Switch,
  Title,
  TrackDetails,
  TrackTitle,
} from "./styles";
import "./styles.css";
import { useVirtualizer } from "@tanstack/react-virtual";

export type PlayQueueProps = {
  previousTracks?: Track[];
  nextTracks?: Track[];
  currentTrack?: Track;
  onPlayTrackAt: (index: number) => void;
  onRemoveTrackAt: (index: number) => void;
};

const PlayQueue: FC<PlayQueueProps> = ({
  previousTracks = [],
  nextTracks = [],
  onPlayTrackAt,
  onRemoveTrackAt,
}) => {
  const theme = useTheme();
  const [active, setActive] = useState("playqueue");
  const parentRef = useRef<HTMLDivElement>(null);
  const { currentIndex, amount } = useMemo(() => {
    return {
      currentIndex: previousTracks.length,
      amount: previousTracks.length + nextTracks.length,
    };
  }, [previousTracks.length, nextTracks.length]);

  // The virtualizer
  const rowVirtualizer = useVirtualizer({
    count: active === "playqueue" ? nextTracks.length : previousTracks.length,
    getScrollElement: () => parentRef.current,
    estimateSize: () => 64,
  });

  const onSwitch = () => {
    if (active === "playqueue") {
      setActive("history");
      return;
    }
    setActive("playqueue");
  };

  const _onPlayTrackAt = (index: number) => {
    if (active === "playqueue") {
      onPlayTrackAt((previousTracks?.length || 0) + index);
      return;
    }
    onPlayTrackAt(index);
  };

  const _onRemoveTrack = (index: number) => {
    if (active === "playqueue") {
      onRemoveTrackAt((previousTracks?.length || 0) + index);
      return;
    }
    onRemoveTrackAt(index);
  };

  const tracks = active === "playqueue" ? nextTracks! : previousTracks!;
  return (
    <Container>
      <Header>
        <Title>{active === "playqueue" ? "Play Queue" : "History"}</Title>
        {amount > 0 && (
          <Title
            style={{ fontSize: 14, color: "#616161", textAlign: "center" }}
          >
            {currentIndex}
            {"  /  "}
            {amount}
          </Title>
        )}
        <Switch onClick={onSwitch}>
          {active === "playqueue" ? "History" : "Play Queue"}
        </Switch>
      </Header>
      <List ref={parentRef}>
        <div
          style={{
            height: rowVirtualizer.getTotalSize()
              ? `${rowVirtualizer.getTotalSize()}px`
              : undefined,
            width: "100%",
            position: "relative",
          }}
        >
          {rowVirtualizer.getVirtualItems().map((virtualItem) => (
            <ListItem
              key={virtualItem.key}
              style={{
                position: "absolute",
                top: 0,
                left: 0,
                width: "calc(100% - 34px)",
                transform: `translateY(${virtualItem.start}px)`,
              }}
            >
              {tracks[virtualItem.index].cover && (
                <div className="album-cover-container queue">
                  <AlbumCover src={tracks[virtualItem.index].cover!} />
                  <div
                    onClick={() => _onPlayTrackAt(virtualItem.index)}
                    className="floating-play queue"
                  >
                    <Play
                      size={16}
                      color={tracks[virtualItem.index].cover ? "#fff" : "#000"}
                    />
                  </div>
                </div>
              )}
              {!tracks[virtualItem.index].cover && (
                <div className="album-cover-container">
                  <AlbumCoverAlt>
                    <TrackIcon width={28} height={28} color="#a4a3a3" />
                  </AlbumCoverAlt>
                  <div
                    onClick={() => _onPlayTrackAt(virtualItem.index)}
                    className="floating-play queue"
                    style={{ left: 19, top: 10 }}
                  >
                    <Play
                      size={16}
                      color={tracks[virtualItem.index].cover ? "#fff" : "#000"}
                    />
                  </div>
                </div>
              )}
              <TrackDetails>
                <TrackTitle>{tracks[virtualItem.index].title}</TrackTitle>
                <Link
                  to={`/artists/${tracks[virtualItem.index].artistId}`}
                  style={{ textDecoration: "none" }}
                >
                  <Artist>{tracks[virtualItem.index].artist}</Artist>
                </Link>
              </TrackDetails>
              <Remove onClick={() => _onRemoveTrack(virtualItem.index)}>
                <CloseOutline size={24} color={theme.colors.text} />
              </Remove>
            </ListItem>
          ))}
          {tracks.length === 0 && active === "playqueue" && (
            <Placeholder>
              No upcoming tracks. Add some to your play queue.
            </Placeholder>
          )}
          {tracks.length === 0 && active === "history" && (
            <Placeholder>
              No history. Play some tracks to see them here.
            </Placeholder>
          )}
        </div>
      </List>
    </Container>
  );
};

export default PlayQueue;
