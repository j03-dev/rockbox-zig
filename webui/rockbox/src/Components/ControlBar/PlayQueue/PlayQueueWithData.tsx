import { FC } from "react";
import PlayQueue from "./PlayQueue";

const PlayQueueWithData: FC = () => {
  return (
    <PlayQueue
      previousTracks={[]}
      nextTracks={[]}
      currentTrack={undefined}
      onPlayTrackAt={() => {}}
      onRemoveTrackAt={() => {}}
    />
  );
};

export default PlayQueueWithData;
