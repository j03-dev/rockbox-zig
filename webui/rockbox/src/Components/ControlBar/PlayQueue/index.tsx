import PlayQueueWithData from "./PlayQueueWithData";

export default PlayQueueWithData;
