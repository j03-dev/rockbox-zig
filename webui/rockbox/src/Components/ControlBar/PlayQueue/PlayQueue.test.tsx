import { render } from "@testing-library/react";
import { vi } from "vitest";
import PlayQueue from "./PlayQueue";
import Providers from "../../../Providers";
import { MockedProvider } from "@apollo/client/testing";
import { mocks } from "../../../mocks";
import { MemoryRouter } from "react-router-dom";
import { nextTracks, previousTracks } from "./mocks";

describe("PlayQueue", () => {
  it("should render", () => {
    const { container } = render(
      <Providers>
        <MemoryRouter initialEntries={["/"]}>
          <MockedProvider mocks={mocks}>
            <PlayQueue
              previousTracks={previousTracks}
              nextTracks={nextTracks}
              currentTrack={undefined}
              onPlayTrackAt={vi.fn()}
              onRemoveTrackAt={vi.fn()}
            />
          </MockedProvider>
        </MemoryRouter>
      </Providers>
    );
    expect(container).toMatchSnapshot();
  });
});
