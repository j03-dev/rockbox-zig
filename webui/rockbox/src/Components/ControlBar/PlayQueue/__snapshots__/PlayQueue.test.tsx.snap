// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlayQueue > should render 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="css-17m5cz6"
    >
      <div
        class="css-o2j9ze"
      >
        <div
          class="css-p7veh6"
        >
          Play Queue
        </div>
        <div
          class="css-p7veh6"
          style="font-size: 14px; color: rgb(97, 97, 97); text-align: center;"
        >
          1
            /  
          4
        </div>
        <div
          class="css-14tdrps"
        >
          History
        </div>
      </div>
      <div
        class="css-1nzjdw3"
      >
        <div
          style="height: 192px; width: 100%; position: relative;"
        />
      </div>
    </div>
  </div>
  <div
    class=""
  />
</div>
`;
